src/entropy.rs

```diff
@@ -0,0 +1,339 @@
+//! Summary statistics (e.g. mean, variance, etc.).
+use ndarray::{Array1, ArrayBase, Data, Dimension};
+use num_traits::Float;
+
+/// Extension trait for `ArrayBase` providing methods
+/// to compute information theory quantities
+/// (e.g. entropy, Kullback–Leibler divergence, etc.).
+pub trait EntropyExt<A, S, D>
+    where
+        S: Data<Elem = A>,
+        D: Dimension,
+{
+    /// Computes the [entropy] *S* of the array values, defined as
+    ///
+    /// ```text
+    ///       n
+    /// S = - ∑ xᵢ ln(xᵢ)
+    ///      i=1
+    /// ```
+    ///
+    /// If the array is empty, `None` is returned.
+    ///
+    /// **Panics** if any element in the array is negative.
+    ///
+    /// ## Remarks
+    ///
+    /// The entropy is a measure used in [Information Theory]
+    /// to describe a probability distribution: it only make sense
+    /// when the array values sum to 1, with each entry between
+    /// 0 and 1 (extremes included).
+    ///
+    /// By definition, *xᵢ ln(xᵢ)* is set to 0 if *xᵢ* is 0.
+    ///
+    /// [entropy]: https://en.wikipedia.org/wiki/Entropy_(information_theory)
+    /// [Information Theory]: https://en.wikipedia.org/wiki/Information_theory
+    fn entropy(&self) -> Option<A>
+    where
+        A: Float;
+
+    /// Computes the [Kullback-Leibler divergence] *Dₖₗ(p,q)* between two arrays,
+    /// where `self`=*p*.
+    ///
+    /// The Kullback-Leibler divergence is defined as:
+    ///
+    /// ```text
+    ///              n
+    /// Dₖₗ(p,q) = - ∑ pᵢ ln(qᵢ/pᵢ)
+    ///             i=1
+    /// ```
+    ///
+    /// If the arrays are empty or their lengths are not equal, `None` is returned.
```

Comment:
I agree that it's not good to combine these two cases. A few options:

* `Result<Option<A>, DimensionMismatch>` would be fine.
* I think `Result<A, ShapeError>` is a bit cleaner (where `ShapeError` is an enum with `Empty` and `ShapeMismatch` variants).
* Another approach is `Option<A>`, returning `None` in the empty case and panicking if `q` cannot be broadcast to the shape of `p`. I generally avoid panicking, but I think this is the best option in this case because:

  * It's more consistent with `ndarray`'s methods that operate on two arrays.
  * If the caller passes in arrays of mismatched shapes, it's clearly a bug in their code.
  * If the arrays have mismatched shapes, it's hard to see how the caller would recover from this error case.

  The question then is why we should return `None` in the empty case instead of panicking. I think this makes sense because it's much more likely than a shape mismatch, doesn't really indicate a bug in the caller, and is more likely to be recoverable.

Reply:
I'd rather go for `Result<A, ShapeError>` - in the end, the caller can choose how to recover from errors (free to call `unwrap` or `expect`).
I can foresee various scenarios for both failure modes:
- empty because you filtered some values out of another array/vec, but now nothing is left;
- shape mismatch, because you read two different samples from files on disk which you expected to have the same number of values but... they don't (real life stories here 😅).

Reply:
To provide a plausible use case: it might be impossible to recover (let the program resume its expected execution flow) but there might be some actions one might want to perform before panicking for example (logging is the first that comes to my mind or sending a request somewhere in a web application context).
One can always catch the panic using stuff like [this](https://doc.rust-lang.org/std/panic/fn.catch_unwind.html) but it's less pleasant and not clear at a first glance at the function signature.

Reply:
Okay, that makes sense. We have to return an `Option`/`Result` anyway, so we might as well handle all the error cases without panicking. Along the same lines, should we return an error on negative values instead of panicking? I'd expect negative values to be a bigger issue than shape mismatches, and we're already checking if values are zero.

Reply:
On that one I am torn, because the cost of checking for negative values scales with the number of elements given that it's an additional check. Do you think it's worth it?

Reply:
Thinking over it again, `Result<Option<A>, DimensionMismatch>` is probably the best signature, given that most of our methods return an `Option<A>` with `None` when arrays are empty. Consistency is probably worth the double wrap (unless we want to use `DimensionMismatch::Empty` instead of `Option` in the rest of the API).

Reply:
I didn't realize earlier that the `ln` of a negative number is `NaN` (except for the noisy float types, which panic). That behavior is fine with me.

Fwiw, I did a quick benchmark of the cost of adding a check for negative numbers, and it's pretty negligible (<2%) (the horizontal axis is `arr.len()`, and vertical axis is time to compute `arr.entropy()`; 'entropy2' is the checked implementation):

![lines](https://user-images.githubusercontent.com/1836850/53456608-0da5f580-39fd-11e9-9c71-8c49d5ce69b8.png)

```rust
    fn entropy2(&self) -> Option<A>
        where
            A: Float
    {
        if self.len() == 0 {
            None
        } else {
            let mut negative = false;
            let entropy = self.mapv(
                |x| {
                    if x < A::zero() {
                        negative = true;
                        A::zero()
                    } else if x == A::zero() {
                        A::zero()
                    } else {
                        x * x.ln()
                    }
                }
            ).sum();
            if negative {
                None
            } else {
                Some(-entropy)
            }
        }
    }
```

(Note that I wouldn't actually return an `Option` in this case; I'd return a `Result` instead.)

Of course, if we check for negative numbers, someone might wonder why we don't check for numbers greater than 1 too.

I could go either way on the negative number checks. The explicit check is nice, but it adds complexity and we aren't checking other things such as values greater than 1 or the sum of values not being 1, so I guess I'd lean towards leaving off negative number checks for right now.

> Thinking over it again, `Result<Option<A>, DimensionMismatch>` is probably the best signature, given that most of our methods return an `Option<A>` with `None` when arrays are empty.

Okay, that's fine.

> unless we want to use `DimensionMismatch::Empty` instead of `Option` in the rest of the API

That wouldn't be too bad, although I don't really like returning an error enum when only one of the variants is possible.